### ISSUE TITLE
Video playback (only if not already similarly requested)

### ISSUE THREAD

**Issue by OpenGreenStreet** [commons-app/apps-android-commons#5983]
### What is the user problem or growth opportunity you want to see solved?

Video playback was requested several times as a basic function.

* see:
* * https://commons.wikimedia.org/w/index.php?title=Commons_talk:Mobile_app&diff=0&oldid=964161967#Proposal_to_put_this_on_the_Main_Page_of_Commons?
* * "If done please include an option like "Please put it on the Main page but only after some key basic features like video playback are added"."

### How do you know that this problem exists today? Why is this important?

Uploaded videos cannot be played from the app.

### Who will benefit from it?

all users

**Comment by parneet-guraya** [comment]
This is interesting. I'd like to work on this if possible. Could you tell me what are the requirements?

**Comment by nicolas-raoul** [comment]
Requirements:
- Ideally see what the Wikipedia app is doing, and do the same.
- Use only open source. To see an example, install the Wikipedia app and open the English "Running" article.
- Keep APK size small, for instance by using Android's readily available components rather than importing new third-party libraries.
- Play most video formats found on Commons. @OpenGreenStreet would you mind listing an example video for each format?
- Play the miniature version of the file, rather than downloading the original which can be huge. See how the Wikipedia app does.
- Not make files lists heavier (for instance Contributions or Explore lists). It is OK if the video is only playable in Media Details.
- Other than that, you can make your own implementation choices. Do not hesitate to ask. 🙂

@OpenGreenStreet Anything important I missed?

**Comment by OpenGreenStreet** [comment]
@nicolas-raoul 

I wrote to [XXX](https://commons.wikimedia.org/wiki/User:Prototyperspective) and asked them to reply instead of me.

I basically just posted his suggestion here




ich habe

**Comment by prototyperspective** [comment]
Thanks for filing this issue and also for pinging me. It's a duplicate of an issue I filed earlier but as long as it gets some momentum behind this issue that's great anyway: #5622 (maybe that issue's title or description was not as good or one could make that linked issue about the other nonsupported filetypes like STL files).

@parneet-guraya it would be amazing if you could look into it; I think one would need to integrate [some open source video player](https://github.com/search?q=video-player+language%3AJava+&type=repositories&s=stars&o=desc) which may not take very long and maybe one could use / copy whatever is used in the Wikipedia app which can play videos (by the way I just noticed it can't make the video fullscreen but otherwise that player seems good and has a jump 10 seconds forward/backward feature).

@nicolas-raoul 
* [This page lists the video formats](https://commons.wikimedia.org/wiki/Special:MediaStatistics#Videos) and one can click on the MIME type column wikilink to see examples
* When playing a video on Commons it also doesn't play the highest resolution version by default; I think the resolution should be configurable in the player / app settings so one can also play videos at low resolution when not in wlan; I think without doing anything it would play that default version and also it streams the file instead of "downloading" it
* Some requirements may not be listed but somewhat self-explanatory; e.g. it shouldn't load very long, have a button to turn the video to fullscreen (maybe even a rotate & fullscreen button), have the skip x seconds forward & backward feature when tapping twice on either side of the video, have a seek feature so you can jump around the video which I think all video players have, and show a thumbnail (ideally the same as on WMC which currently still can't be configured)

**Comment by OpenGreenStreet** [comment]
@prototyperspective 
Thank you for your feedback.
I had already thought about integrating an open source video player. Would VLC perhaps be an option?
* https://stackoverflow.com/questions/22073970/how-to-embed-vlc-media-player-to-my-android-app
* https://www.videolan.org/vlc/libvlc.html

**Comment by parneet-guraya** [comment]
Thankyou @nicolas-raoul & @prototyperspective for specifying the requirements, I will pick this issue up shortly.


As far as player is concerned we have an official library from google [Jetpack Media3](https://developer.android.com/media/media3). It supports both audio,video & images (never used it for that but can be useful for `gif` ) . I already have worked with it in the past a little example ([migration](https://github.com/owncloud/android/pull/4177)), you can see how interface would look like. It has all the basic functionality one can expect from a player; play/pause , slide to seek/seek fw/bw & we can have full screen enter/exit as well.

Lastly, loading time or stream quality would depend on if API provide us with adaptive streaming type.

Thanks again :-)

**Comment by sivaraam** [comment]
Just to confirm. Are you still looking into this @parneet-guraya ?

**Comment by prototyperspective** [comment]
This is one of the top issues with the app, maybe the most important and if it gets implemented audio (#5915) will probably be playable as well. parneet-guraya doesn't seem to be working on this, would be great if somebody could pick it up. For example, probably https://github.com/mpv-android/mpv-android https://github.com/anilbeesetti/NextPlayer or https://github.com/KhubaibKhan4/MediaPlayer-KMP (all in Kotlin) could be used.

**Comment by Amdrel** [comment]
Google's Media3 player seems to support everything that's needed to play Commons supported formats and subtitles, etc. I'm not sure what the other 3rd party players offer that we need that Media3 doesn't already provide.

I can pick up this ticket and give it a go.

**Update**: I've made good progress on this. I've got something that works and I'm just polishing things up and seeing if I can write some useful test cases for it.
